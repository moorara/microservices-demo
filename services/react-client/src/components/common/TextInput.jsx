import React from 'react'
import PropTypes from 'prop-types'

const TextInput = ({ name, label, placeholder, value, isOk, isError, help, onChange }) => {
  let inputClass = 'input'
  let helpClass = 'help'

  if (isOk) {
    inputClass += ' is-success'
    helpClass += ' is-success'
  } else if (isError) {
    inputClass += ' is-danger'
    helpClass += ' is-danger'
  }

  return (
    <div className="field">
      <label className="label" htmlFor={name}>{label}</label>
      <div className="control has-icons-right">
        <input
          type="text"
          className={inputClass}
          name={name}
          placeholder={placeholder}
          value={value}
          onChange={onChange} />
        { isOk &&
          <span className="icon is-small is-right">
            <i className="fa fa-check-circle" />
          </span>
        }
        { isError &&
          <span className="icon is-small is-right">
            <i className="fa fa-exclamation-circle" />
          </span>
        }
      </div>
      { help && <p className={helpClass}>{help}</p> }
    </div>
  )
}

TextInput.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.string,
  isOk: PropTypes.bool,
  isError: PropTypes.bool,
  help: PropTypes.string,
  onChange: PropTypes.func
}

TextInput.defaultProps = {
  placeholder: '',
  value: '',
  isOk: false,
  isError: false,
  help: ''
}

export default TextInput
