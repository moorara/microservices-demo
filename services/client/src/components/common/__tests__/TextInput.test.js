import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import TextInput from '../TextInput'

describe('TextInput', () => {
  let handleChange

  beforeEach(() => {
    // jest.fn(...) is a shorthand for jest.fn().mockImplementation(...)
    handleChange = jest.fn(e => {})
  })

  test('renders react component with placeholder', () => {
    const wrapper = shallow(<TextInput name="firstName" label="First Name" placeholder="Your Name" onChange={handleChange} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
    expect(wrapper.find('.label').text()).toBe('First Name')
    expect(wrapper.find('div.control').exists()).toBeTruthy()
    expect(wrapper.find('span.icon').exists()).toBeFalsy()
    expect(wrapper.find('p.help').exists()).toBeFalsy()
    expect(wrapper.find('input').find({ type: 'text', name: 'firstName', placeholder: 'Your Name', value: '' }).exists()).toBeTruthy()

    wrapper.find('input').find({ type: 'text' }).simulate('change', 'my name')
    expect(handleChange).toHaveBeenCalledWith('my name')
  })
  test('renders react component with value', () => {
    const wrapper = shallow(<TextInput name="firstName" label="First Name" value="Me" onChange={handleChange} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
    expect(wrapper.find('.label').text()).toBe('First Name')
    expect(wrapper.find('div.control').exists()).toBeTruthy()
    expect(wrapper.find('span.icon').exists()).toBeFalsy()
    expect(wrapper.find('p.help').exists()).toBeFalsy()
    expect(wrapper.find('input').find({ type: 'text', name: 'firstName', placeholder: '', value: 'Me' }).exists()).toBeTruthy()

    wrapper.find('input').find({ type: 'text' }).simulate('change', 'my name')
    expect(handleChange).toHaveBeenCalledWith('my name')
  })
  test('renders react component in error case', () => {
    const wrapper = shallow(<TextInput name="firstName" label="First Name" isError help="name is invalid" onChange={handleChange} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
    expect(wrapper.find('.label').text()).toBe('First Name')
    expect(wrapper.find('div.control').exists()).toBeTruthy()
    expect(wrapper.find('span.icon i').prop('className')).toContain('fa-exclamation-circle')
    expect(wrapper.find('p.help').text()).toBe('name is invalid')
    expect(wrapper.find('input').find({ type: 'text', name: 'firstName', placeholder: '', value: '' }).exists()).toBeTruthy()

    wrapper.find('input').find({ type: 'text' }).simulate('change', 'my name')
    expect(handleChange).toHaveBeenCalledWith('my name')
  })
  test('renders react component in success case', () => {
    const wrapper = shallow(<TextInput name="firstName" label="First Name" isOk help="name is valid" onChange={handleChange} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
    expect(wrapper.find('.label').text()).toBe('First Name')
    expect(wrapper.find('div.control').exists()).toBeTruthy()
    expect(wrapper.find('span.icon i').prop('className')).toContain('fa-check-circle')
    expect(wrapper.find('p.help').text()).toBe('name is valid')
    expect(wrapper.find('input').find({ type: 'text', name: 'firstName', placeholder: '', value: '' }).exists()).toBeTruthy()

    wrapper.find('input').find({ type: 'text' }).simulate('change', 'my name')
    expect(handleChange).toHaveBeenCalledWith('my name')
  })
})
