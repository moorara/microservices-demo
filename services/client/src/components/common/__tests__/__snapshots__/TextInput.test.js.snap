// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders react component in error case 1`] = `
<div
  className="field"
>
  <label
    className="label"
    htmlFor="firstName"
  >
    First Name
  </label>
  <div
    className="control has-icons-right"
  >
    <input
      className="input is-danger"
      name="firstName"
      onChange={[MockFunction]}
      placeholder=""
      type="text"
      value=""
    />
    <span
      className="icon is-small is-right"
    >
      <i
        className="fa fa-exclamation-circle"
      />
    </span>
  </div>
  <p
    className="help is-danger"
  >
    name is invalid
  </p>
</div>
`;

exports[`TextInput renders react component in success case 1`] = `
<div
  className="field"
>
  <label
    className="label"
    htmlFor="firstName"
  >
    First Name
  </label>
  <div
    className="control has-icons-right"
  >
    <input
      className="input is-success"
      name="firstName"
      onChange={[MockFunction]}
      placeholder=""
      type="text"
      value=""
    />
    <span
      className="icon is-small is-right"
    >
      <i
        className="fa fa-check-circle"
      />
    </span>
  </div>
  <p
    className="help is-success"
  >
    name is valid
  </p>
</div>
`;

exports[`TextInput renders react component with placeholder 1`] = `
<div
  className="field"
>
  <label
    className="label"
    htmlFor="firstName"
  >
    First Name
  </label>
  <div
    className="control has-icons-right"
  >
    <input
      className="input"
      name="firstName"
      onChange={[MockFunction]}
      placeholder="Your Name"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`TextInput renders react component with value 1`] = `
<div
  className="field"
>
  <label
    className="label"
    htmlFor="firstName"
  >
    First Name
  </label>
  <div
    className="control has-icons-right"
  >
    <input
      className="input"
      name="firstName"
      onChange={[MockFunction]}
      placeholder=""
      type="text"
      value="Me"
    />
  </div>
</div>
`;
